dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/InactivityValidationAwareConnectionKeepAliveStrategy.java

```diff
@@ -0,0 +1,88 @@
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HeaderElements;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.hc.core5.util.TimeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class InactivityValidationAwareConnectionKeepAliveStrategy implements ConnectionKeepAliveStrategy {
+    private static final Logger log =
+            LoggerFactory.getLogger(InactivityValidationAwareConnectionKeepAliveStrategy.class);
+
+    private static final ConnectionKeepAliveStrategy DELEGATE = DefaultConnectionKeepAliveStrategy.INSTANCE;
+    private final PoolingHttpClientConnectionManager connectionManager;
+    private final String clientName;
+    private final TimeValue defaultValidateAfterInactivity;
+    /**
+     * This field is used for observability. It's possible, though unlikely, that the value can get out of sync
+     * with the connection manager in some scenarios.
+     */
+    private final AtomicReference<TimeValue> currentValidationInterval;
+
+    InactivityValidationAwareConnectionKeepAliveStrategy(
+            PoolingHttpClientConnectionManager connectionManager, String clientName) {
+        this.connectionManager = connectionManager;
+        this.clientName = clientName;
+        // Store the initial inactivity interval to restore if responses re received without
+        // keep-alive headers.
+        this.defaultValidateAfterInactivity = connectionManager.getValidateAfterInactivity();
+        this.currentValidationInterval = new AtomicReference<>(defaultValidateAfterInactivity);
+    }
+
+    @Override
+    public TimeValue getKeepAliveDuration(HttpResponse response, HttpContext context) {
+        TimeValue result = DELEGATE.getKeepAliveDuration(response, context);
+        if (result != null
+                // Only use keep-alive values from 2xx responses
+                && response.getCode() / 100 == 2) {
+            TimeValue newInterval =
+                    containsKeepAliveHeaderWithTimeout(response) ? result : defaultValidateAfterInactivity;
+            TimeValue previousInterval = currentValidationInterval.getAndSet(newInterval);
+            if (!Objects.equals(previousInterval, newInterval)) {
+                log.info(
+                        "Updating the validation interval for {} from {} to {}",
+                        SafeArg.of("client", clientName),
+                        SafeArg.of("previousInterval", previousInterval),
+                        SafeArg.of("newInterval", newInterval));
+            }
+            // Simple volatile write, no need to protect this in the getAndSet check. The getAndSet may race this call
+            // so it's best to completely decouple the two.
+            connectionManager.setValidateAfterInactivity(newInterval);
```

Comment:
Do we want to do any sanity checking of the range here? like what happens if some bizarre proxy/server sends timeout=-1, timeout=999999 etc?

Reply:
non-positive values should be omitted since those values as timeouts are unclear, some components use negative and zero values as infinite. If a server wants to advertise an exceedingly long timeout, I'm inclined to trust it for now. We could add a maximum later. In the worst case, when a server goes down, our retries will get us into a good state.